### PR TITLE
UCP: Add atomics support, and allow passing "features" to UCP

### LINKED ISSUE #242: UCP: add RMA protocols
- zcopy, put, get, amo

### LINKED ISSUE #262: UCP RMA should support transports without AM

### LINKED ISSUE #280: transport retry exceeded with ucx_perftest using ucp
```
mpirun -n 2 --map-by node -report-bindings -map-by dist:span -mca rmaps_dist_device mlx5_1  --display-map -x UCX_TLS=cm,rc_mlx5 -x UCX_DEVICES=mlx5_1:1 ucx_perftest -t ucp_put_bw -b <>/ucx/share/ucx/perftest/msg_pow2 -f -v
Data for JOB [376,1] offset 0

========================   JOB MAP   ========================

Data for node: thor031 Num slots: 1    Max slots: 0    Num procs: 1
        Process OMPI jobid: [376,1] App: 0 Process rank: 0

Data for node: thor032 Num slots: 1    Max slots: 0    Num procs: 1
        Process OMPI jobid: [376,1] App: 0 Process rank: 1

=============================================================
[thor031:88043] MCW rank 0 bound to socket 0[core 0[hwt 0]]: [B/././././././././././././.][./././././././././././././.]
[thor032:94138] MCW rank 1 bound to socket 0[core 0[hwt 0]]: [B/././././././././././././.][./././././././././././././.]
[warn] Epoll ADD(1) on fd 0 failed.  Old events were 0; read change was 1 (add); write change was 0 (none): Operation not permitted
msg_pow2,iterations,typical_lat,avg_lat,overall_lat,avg_bw,overall_bw,avg_mr,overall_mr
1,2000000,0.070,0.114,0.114,8.34,8.34,8748702,8748702
2,2000000,0.070,0.114,0.114,16.69,16.69,8749535,8749535
4,2000000,0.070,0.114,0.114,33.37,33.37,8748579,8748579
8,2000000,0.070,0.078,0.078,97.90,97.90,12832496,12832496
12,2000000,0.070,0.105,0.105,127.12,127.12,9521428,9521428
16,2000000,0.070,0.114,0.114,133.48,133.48,8747939,8747939
24,2000000,0.070,0.074,0.074,309.75,309.75,13533001,13533001
32,2000000,0.088,0.091,0.091,334.48,334.48,10960368,10960368
40,2000000,0.088,0.088,0.088,432.53,432.53,11338469,11338469
48,2000000,0.088,0.088,0.088,519.18,519.18,11341666,11341666
64,2000000,0.088,0.088,0.088,691.91,691.91,11336267,11336267
80,2000000,0.088,0.088,0.088,864.54,864.54,11331662,11331662
96,2000000,0.088,0.088,0.088,1038.20,1038.20,11339861,11339861
128,1400000,0.088,0.088,0.088,1384.55,1384.55,11342260,11342260
256,700000,0.180,0.181,0.181,1348.32,1348.32,5522726,5522726
300,700000,0.181,0.181,0.181,1580.32,1580.32,5523617,5523617
512,300000,0.279,0.285,0.285,1710.36,1710.36,3502827,3502827
1024,200000,0.480,0.483,0.483,2021.18,2021.18,2069693,2069693
2048,100000,0.167,0.174,0.174,11249.27,11249.27,5759624,5759624
3000,100000,0.250,0.253,0.253,11317.28,11317.28,3955676,3955676
4096,100000,0.334,0.339,0.339,11515.96,11515.96,2948087,2948087
6000,100000,0.493,0.495,0.495,11569.62,11569.62,2021937,2021937
8192,80000,0.668,0.668,0.668,11694.07,11694.07,1496841,1496841
10000,80000,0.812,0.828,0.828,11518.21,11518.21,1207772,1207772
16384,40000,1.330,1.338,1.338,11680.51,11680.51,747552,747552
[1435278370.403175] [thor031:88053:0]    ib_mlx5_log.c:106  UCX  ERROR Error on QP 0x4fb wqe[22463]: Transport retry count exceeded (synd 0x15 vend 0x81) opcode RDMA_WRITE
25000,40000,2.042,2.096,2.096,11377.30,11377.30,477199,477199
32768,20000,2.668,2.677,2.677,11673.62,11673.62,373556,373556
45000,20000,3.653,3.684,3.684,11649.13,11649.13,271444,271444
65536,10000,5.338,5.352,5.352,11677.68,11677.68,186843,186843
100000,10000,8.141,8.156,8.156,11692.26,11692.26,122602,122602
131072,5000,10.724,10.719,10.719,11661.62,11661.62,93293,93293
262144,2500,21.428,21.430,21.430,11666.03,11666.03,46664,46664
524288,1200,42.742,42.837,42.837,11672.13,11672.13,23344,23344
[1435278388.656539] [thor031:88053:0]    ib_mlx5_log.c:106  UCX  ERROR Error on QP 0x50b wqe[18940]: Transport retry count exceeded (synd 0x15 vend 0x81) opcode RDMA_WRITE
[1435278388.656655] [thor031:88053:0]    ib_mlx5_log.c:106  UCX  ERROR Error on QP 0x50b wqe[18941]: WR flushed because QP in error state (synd 0x5 vend 0xf9) opcode RDMA_WRITE
[1435278388.656662] [thor031:88053:0]    ib_mlx5_log.c:106  UCX  ERROR Error on QP 0x50b wqe[18942]: WR flushed because QP in error state (synd 0x5 vend 0xf9) opcode RDMA_WRITE
[1435278388.656670] [thor031:88053:0]    ib_mlx5_log.c:106  UCX  ERROR Error on QP 0x50b wqe[18943]: WR flushed because QP in error state (synd 0x5 vend 0xf9) opcode RDMA_WRITE
```

### LINKED ISSUE #296: Add ucp perftest to gtest envelope

### PR DESCRIPTION
**Pull request by yosefe**
fixes #242 
fixes #262 
fixes #280
fixes #296

### CODE REVIEW

File: src/ucp/api/ucp.h [hunk 1]
```diff
@@ -15,6 +15,17 @@
 
 /**
  * @ingroup CONTEXT
+ * @brief UCP features
+ */
+enum {
+    UCP_FEATURE_TAG = UCS_BIT(1),  /* Request tag matching support */
+    UCP_FEATURE_RMA = UCS_BIT(2),  /* Request remote memory access support */
+    UCP_FEATURE_AMO = UCS_BIT(3)   /* Request atomic operations support */
+};
```
Comment (shamisp) [inline src/ucp/api/ucp.h]: Please use doxygen format; /**< XXX */

**Comment by yosefe** [comment]
bot:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com:8888/jenkins/job/gh-ucx-pr/708/ for details.

File: src/ucp/api/ucp.h [hunk 2]
```diff
@@ -99,14 +119,16 @@ void ucp_config_print(const ucp_config_t *config, FILE *stream,
...
+ *                                 features would result in undefined behavior.
  * @param [in]  request_headroom  How many bytes to reserve before every request
  *                                 returned by non-blocking operations.
  * @param [out] context_p         Filled with a ucp context handle.
  *
  * @return Error code.
  */
-ucs_status_t ucp_init(const ucp_config_t *config, size_t request_headroom,
-                      ucp_context_h *context_p);
+ucs_status_t ucp_init(unsigned features, size_t request_headroom,
```
Comment (shamisp) [inline src/ucp/api/ucp.h]: let'b be more specific and allocate 32 or 64 bit for this. uint64_t ?

**Comment by shamisp** [comment]
See my comment about size of the feature flag. Otherwise, it is fine.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com:8888/jenkins/job/gh-ucx-pr/724/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/161/ for details. (Mellanox internal link)

**Comment by yosefe** [comment]
tests in the lab are failing for now because of debian build, i think we can merge it

**Comment by shamisp** [comment]
:+1:

**Pull request merged**